File: src/data_structures/task.rs
use super::punctuation::Punctuation;
use super::term::Term;
use super::truth_value::TruthValue;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::fmt;
use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
use std::sync::Arc;

/// Represents a task in the SeNARS system.
///
/// `priority` and `accessed_at` use atomic types for efficient, lock-free updates,
/// allowing these fields to be modified even when the Task is behind an `Arc`.
#[derive(Clone)]
pub struct Task {
    pub term: Arc<Term>,
    pub punctuation: Punctuation,
    pub truth: Option<TruthValue>,
    /// The priority of the task, stored as the bits of an f32.
    priority: AtomicU32,
    /// The timestamp of the last access.
    accessed_at: AtomicU64,
    pub created_at: u64,
    pub occurrence_time: Option<u64>,
    pub expiration_time: Option<u64>,
    pub derivation_path: Option<Vec<String>>,
}

impl Task {
    pub fn new(
        term: Arc<Term>,
        punctuation: Punctuation,
        truth: Option<TruthValue>,
        created_at: u64,
        occurrence_time: u64,
    ) -> Self {
        Task {
            term,
            punctuation,
            truth,
            priority: AtomicU32::new(0.5f32.to_bits()), // Default priority
            accessed_at: AtomicU64::new(created_at),
            created_at,
            occurrence_time: Some(occurrence_time),
            expiration_time: None,
            derivation_path: None,
        }
    }

    pub fn term(&self) -> &Arc<Term> {
        &self.term
    }

    pub fn get_priority(&self) -> f32 {
        f32::from_bits(self.priority.load(Ordering::Relaxed))
    }

    pub fn set_priority(&self, new_priority: f32) {
        self.priority.store(new_priority.to_bits(), Ordering::Relaxed);
    }

    pub fn get_accessed_at(&self) -> u64 {
        self.accessed_at.load(Ordering::Relaxed)
    }

    pub fn set_accessed_at(&self, new_time: u64) {
        self.accessed_at.store(new_time, Ordering::Relaxed);
    }

    pub fn is_belief(&self) -> bool {
        self.punctuation == Punctuation::Belief
    }

    pub fn is_goal(&self) -> bool {
        self.punctuation == Punctuation::Goal
    }

    pub fn is_question(&self) -> bool {
        self.punctuation == Punctuation::Question
    }

    pub fn is_expired(&self, current_time: u64) -> bool {
        if let Some(expiration) = self.expiration_time {
            current_time > expiration
        } else {
            false
        }
    }
}

// Manual trait implementations due to atomic fields.

impl fmt::Debug for Task {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Task")
            .field("term", &self.term)
            .field("punctuation", &self.punctuation)
            .field("truth", &self.truth)
            .field("priority", &self.get_priority())
            .field("accessed_at", &self.get_accessed_at())
            .field("created_at", &self.created_at)
            .field("occurrence_time", &self.occurrence_time)
            .field("expiration_time", &self.expiration_time)
            .field("derivation_path", &self.derivation_path)
            .finish()
    }
}

impl PartialEq for Task {
    fn eq(&self, other: &Self) -> bool {
        self.term == other.term
            && self.punctuation == other.punctuation
            && self.truth == other.truth
            && self.get_priority() == other.get_priority()
            && self.get_accessed_at() == other.get_accessed_at()
            && self.created_at == other.created_at
            && self.occurrence_time == other.occurrence_time
            && self.expiration_time == other.expiration_time
            && self.derivation_path == other.derivation_path
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let term_str = self.term().to_string();
        let punc_str = self.punctuation.to_string();
        match &self.truth {
            Some(truth) => write!(f, "{}{} {}", term_str, punc_str, truth),
            None => write!(f, "{}{}", term_str, punc_str),
        }
    }
}

// --- Serialization ---

#[derive(Serialize, Deserialize)]
struct TaskSerdeHelper {
    #[serde(with = "super::term::arc_term_serde")]
    term: Arc<Term>,
    punctuation: Punctuation,
    truth: Option<TruthValue>,
    priority: f32,
    accessed_at: u64,
    created_at: u64,
    occurrence_time: Option<u64>,
    expiration_time: Option<u64>,
    derivation_path: Option<Vec<String>>,
}

impl Serialize for Task {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let helper = TaskSerdeHelper {
            term: self.term.clone(),
            punctuation: self.punctuation,
            truth: self.truth.clone(),
            priority: self.get_priority(),
            accessed_at: self.get_accessed_at(),
            created_at: self.created_at,
            occurrence_time: self.occurrence_time,
            expiration_time: self.expiration_time,
            derivation_path: self.derivation_path.clone(),
        };
        helper.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for Task {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let helper = TaskSerdeHelper::deserialize(deserializer)?;
        Ok(Task {
            term: helper.term,
            punctuation: helper.punctuation,
            truth: helper.truth,
            priority: AtomicU32::new(helper.priority.to_bits()),
            accessed_at: AtomicU64::new(helper.accessed_at),
            created_at: helper.created_at,
            occurrence_time: helper.occurrence_time,
            expiration_time: helper.expiration_time,
            derivation_path: helper.derivation_path,
        })
    }
}

File: core/data_structures/Task.js
import { Punctuation, punctuationEquals } from './Punctuation.js';
import { TruthValue } from './TruthValue.js';

// Task class - represents a task in the SeNARS system
export class Task {
  constructor({
    term,
    punctuation = Punctuation.Belief,
    truth = null,
    priority = 0.5,
    accessedAt = Date.now(),
    createdAt = Date.now(),
    occurrenceTime = null,
    expirationTime = null,
    derivationPath = null
  } = {}) {
    this.term = term;
    this.punctuation = punctuation;
    this.truth = truth;
    this.priority = priority;
    this.accessedAt = accessedAt;
    this.createdAt = createdAt;
    this.occurrenceTime = occurrenceTime;
    this.expirationTime = expirationTime;
    this.derivationPath = derivationPath;
  }

  // Factory method to create a new task
  static new(term, punctuation, truth, createdAt, occurrenceTime) {
    return new Task({
      term,
      punctuation,
      truth,
      priority: 0.5, // Default priority
      accessedAt: createdAt,
      createdAt,
      occurrenceTime: occurrenceTime !== undefined ? occurrenceTime : null,
      expirationTime: null,
      derivationPath: null
    });
  }

  // Getters for convenience
  getTerm() {
    return this.term;
  }

  getPriority() {
    return this.priority;
  }

  setPriority(newPriority) {
    this.priority = Math.max(0, Math.min(1, newPriority)); // Clamp between 0 and 1
  }

  getAccessedAt() {
    return this.accessedAt;
  }

  setAccessedAt(newTime) {
    this.accessedAt = newTime;
  }

  isBelief() {
    return this.punctuation === Punctuation.Belief;
  }

  isGoal() {
    return this.punctuation === Punctuation.Goal;
  }

  isQuestion() {
    return this.punctuation === Punctuation.Question;
  }

  isExpired(currentTime = Date.now()) {
    if (this.expirationTime !== null) {
      return currentTime > this.expirationTime;
    }
    return false;
  }

  // Override toString for Narsese representation
  toString() {
    const termStr = this.term.toString();
    const puncStr = this.punctuation;
    if (this.truth) {
      return `${termStr}${puncStr} ${this.truth}`;
    }
    return `${termStr}${puncStr}`;
  }

  // Override valueOf and toJSON for proper serialization
  valueOf() {
    return this.toString();
  }

  toJSON() {
    return {
      term: this.term,
      punctuation: this.punctuation,
      truth: this.truth,
      priority: this.priority,
      accessedAt: this.accessedAt,
      createdAt: this.createdAt,
      occurrenceTime: this.occurrenceTime,
      expirationTime: this.expirationTime,
      derivationPath: this.derivationPath
    };
  }

  // Check equality based on multiple properties (optimized)
  equals(other) {
    if (!(other instanceof Task)) return false;
    // Fast path: if terms have different hashes, they're not equal
    if (this.term.hash !== other.term.hash) return false;
    
    // Compare primitive properties first (fast)
    if (this.punctuation !== other.punctuation ||
        this.priority !== other.priority ||
        this.accessedAt !== other.accessedAt ||
        this.createdAt !== other.createdAt ||
        this.occurrenceTime !== other.occurrenceTime ||
        this.expirationTime !== other.expirationTime) {
      return false;
    }
    
    // Compare truth values
    if (this.truth !== other.truth) {
      if (this.truth === null || other.truth === null || 
          !this.truth.equals || !other.truth.equals ||
          !this.truth.equals(other.truth)) {
        return false;
      }
    }
    
    // Compare derivation paths efficiently
    if (this.derivationPath !== other.derivationPath) {
      if (this.derivationPath === null || other.derivationPath === null) return false;
      if (this.derivationPath.length !== other.derivationPath.length) return false;
      for (let i = 0; i < this.derivationPath.length; i++) {
        if (this.derivationPath[i] !== other.derivationPath[i]) return false;
      }
    }
    
    return true;
  }

  // Strict equality for debugging purposes
  deepEquals(other) {
    return this.equals(other);
  }
}